### DISCUSSION TITLE
Examples for UAC2 with multiple microphones (array) and multiple sampling rates

### DISCUSSION THREAD

**Discussion by doumdi** [hathach/tinyusb#848] [Q&A]
Hi,

I would like to know if it would be technically possible with tinyusb to implement a multi-channel sound card that would be useful for beamforming applications. 

I was thinking implementing this with RP2040 PIO using up to 16 daisy-chained TDM microphones @ 16kHz. Of course, the maximum sampling rate would depend on the number of channels fitting in a 1ms USB frame. Using TDM microphones would be fairly straightforward compared to PDM microphones that need to be filtered. I am thinking using TDK/InvenSense ICS-52000 microphones that would require a clock rate of 512 x 16000 (8.192 MHz) for 16 microphones.

Adding an example for this kind of application would be awesome. 

Any thoughts ?

Best regards,

Dominic

**Comment by hathach** [comment]
Yeah, adding an example like this would be very awesome. There is a 4 channel mic example that you could take a look at https://github.com/hathach/tinyusb/tree/master/examples/device/audio_4_channel_mic . The audio driver is still WIP, and by its nature is kind of complicated at the moment. In the long run, I will try to see if we could make it simpler to use.

**Comment by PanRe** [comment]
I am doing something like you proposed too. From an USB point of view, there is no problem, you can expand my example of 4 channels easily to 16. However, i think the theoretical maximum of the standard Windows UAC2 driver only supports up to 8 channels. So you may need to write your own host USB driver. I don't know the limits of Linux drivers... Let me know how you are doing!
